algosdk/dryrun_results.py

```diff
@@ -0,0 +1,227 @@
+                    truncate(stack, spc.max_width),
+                ]
+            )
+
+        return (
+            tabulate(
+                lines,
+                headers,
+                disable_numparse=True,
+                tablefmt=TableFormat(
+                    headerrow=DataRow("", " |", ""),
+                    datarow=DataRow("", " |", ""),
+                    padding=0,
+                    lineabove=None,
+                    linebelowheader=None,
+                    linebetweenrows=None,
+                    linebelow=None,
+                    with_header_hide=None,
+                ),
+            )
+            + "\n"
+        )
+
+    def app_trace(self, spc: StackPrinterConfig = None) -> str:
+        if not hasattr(self, "app_call_trace"):
+            return ""
+
+        if spc == None:
+            spc = StackPrinterConfig(top_of_stack_first=False)
+
+        return self.trace(self.app_call_trace, self.disassembly, spc=spc)
+
+    def lsig_trace(self, spc: StackPrinterConfig = None) -> str:
+        if not hasattr(self, "logic_sig_trace"):
+            return ""
+
+        if (
+            not hasattr(self, "logic_sig_disassembly")
+            or self.logic_sig_disassembly is None
+        ):
+            return ""
+
+        if spc == None:
+            spc = StackPrinterConfig(top_of_stack_first=False)
+
+        return self.trace(
+            self.logic_sig_trace, self.logic_sig_disassembly, spaces=spc
+        )
+
+
+class DryrunTrace:
+    def __init__(self, trace: List[dict]):
+        self.trace = [DryrunTraceLine(line) for line in trace]
+
+    def get_trace(self) -> List[str]:
+        return [line.trace_line() for line in self.trace]
+
+
+class DryrunTraceLine:
+    def __init__(self, tl):
+        self.line = tl["line"]
+        self.pc = tl["pc"]
+
+        self.error = ""
+        if "error" in tl:
+            self.error = tl["error"]
+
+        self.scratch = []
+        if "scratch" in tl:
+            self.scratch = [DryrunStackValue(sv) for sv in tl["scratch"]]
+
+        self.stack = [DryrunStackValue(sv) for sv in tl["stack"]]
+
+
+class DryrunStackValue:
+    def __init__(self, v):
+        self.type = v["type"]
+        self.bytes = v["bytes"]
+        self.int = v["uint"]
+
+    def __str__(self) -> str:
+        if len(self.bytes) > 0:
+            return "0x" + base64.b64decode(self.bytes).hex()
+        return str(self.int)
+
+    def __eq__(self, other: "DryrunStackValue"):
+        return (
+            self.type == other.type
+            and self.bytes == other.bytes
+            and self.int == other.int
+        )
+
+
+def truncate(s: str, max_width: int) -> str:
+    if len(s) > max_width and max_width > 0:
+        return s[:max_width] + "..."
+    return s
+
+
+def scratch_to_string(
+    prev_scratch: List[DryrunStackValue], curr_scratch: List[DryrunStackValue]
+) -> str:
+    if len(curr_scratch) == 0:
+        return ""
```

Comment:
```suggestion
    if not curr_scratch:
        return ""
```
slightlhy more pythoni